Catch typesafeconfig exceptions

melrief opened the pull request:
Solves #148 

This PR is an attempt to catch and wrap inside failures the exception thrown by the typesafe config parser and loader. Changes:

- add `ConfigFactoryWrapper` as safer version of the typesafe `ConfigFactory`
- change the `pureconfig.load*` methods to use the `ConfigFactoryWrapper` methods instead of the `ConfigFactory` ones
- change `pureconfig.loadConfigFromFiles` to work on `Path` instead of `File`. Not sure why I accepted a PR working on something different but we are using `Path` everywhere and I'd like to be consistent

p.s. I've noticed in this and another PR that there are some changes on the styles of the files. Something is not working properly but I'm not sure what. I'll leave the style changes in the PR because we should have well formatted code...

melrief committed:
catch typesafe config exceptions in pureconfig.load* methods

melrief committed:
add issue link to the CHANGELOG

melrief committed:
fix code for scala <2.12

leifwickland commented on core/src/main/scala/pureconfig/package.scala:
```diff
@@ -252,15 +258,18 @@ package object pureconfig {
    * @param files Files ordered in decreasing priority containing part or all of a `Config`. Must not be empty.
    */
-  def loadConfigFromFiles[Config: ConfigConvert](files: Traversable[java.io.File]): Either[ConfigReaderFailures, Config] = {
+  def loadConfigFromFiles[Config: ConfigConvert](files: Traversable[Path]): Either[ConfigReaderFailures, Config] = {
```
Sorry. This is my fault originally.

leifwickland commented on core/src/main/scala/pureconfig/backend/TypesafeConfigWrapper.scala:
```diff
@@ -0,0 +1,38 @@
+  private def unsafeToEither[A](f: => A): Either[ConfigReaderFailures, A] = {
+    try (Right(f)) catch {
+      case NonFatal(e) =>
+        Left(ConfigReaderFailures(ThrowableFailure(e, None)))
```
Thanks for doing this.

jcazevedo replied:
I think we can do a little better here. If we catch [`ConfigException`](https://typesafehub.github.io/config/latest/api/com/typesafe/config/ConfigException.html), we have access to a `ConfigOrigin` which we can then use to create a `ConfigValueLocation`. This would add more context to our errors. Also, we can map some types of `ConfigException` (I see at least [`ConfigException.Parse`](https://typesafehub.github.io/config/latest/api/com/typesafe/config/ConfigException.Parse.html) and [`ConfigException.NotResolved`](https://typesafehub.github.io/config/latest/api/com/typesafe/config/ConfigException.NotResolved.html)) which can be mapped to more semantically rich failures of PureConfig, other than just a `ThrowableFailure`.

leifwickland commented on core/src/main/scala/pureconfig/package.scala:
```diff
@@ -23,8 +24,8 @@ package object pureconfig {
   def loadConfig[Config](implicit conv: ConfigConvert[Config]): Either[ConfigReaderFailures, Config] = {
-    ConfigFactory.invalidateCaches()
-    loadConfig[Config](ConfigFactory.load())(conv)
+    invalidateCaches()
```
If this fails the method will continue loading the configuration mixed with the stale cache. 
Previously if this failed by throwing an exception, the method would fail.
I don' think you intended to change the behavior.
It seems like we need to chain `invalidateCaches()` with the following load call to propagate errors correctly.
You roughly want the `*>` from scalaz. It might be worth implementing something like that here for concision.
Ditto multiple places below.

melrief replied:
What a mistake! I've done this very late and I missed this obvious mistake.

leifwickland reviewed (changes requested):
I don't think you intended the behavior change around `invalidateCache`

melrief committed:
propagate invalidateCaches errors

melrief commented:
@leifwickland thanks for the review! I've addressed the issue with `invalidateCaches`, let me know what do you think.

leifwickland commented on core/src/main/scala/pureconfig/package.scala:
```diff
@@ -143,9 +166,13 @@ package object pureconfig {
+        rawConfig <- loadFile(path).right
+        config <- loadConfig[Config](rawConfig)(conv).right
+      } yield config
+    getResultOrThrow[Config](errorOrConfig)
```
Do you recall why we didn't implement this as:
```
getResultOrThrow[Config](loadConfig(path)(conv))
```

Ditto for the other `loadConfigOrThrow` methods.

melrief replied:
No but it's equivalent, let me fix it

melrief replied:
Fixed with 612b620

leifwickland replied:
Thanks!
It seems like we'd want to reuse the `loadConfig` implementations for for the other `loadConfigOrThrow` methods.

melrief replied:
I was thinking the same and I was wondering why we didn't implemented the `loadConfigOrThrow` by using `loadConfig`. Then I noticed that the location is filled differently, e.g.

```scala
  def loadConfig[Config](namespace: String)(implicit conv: ConfigConvert[Config]): Either[ConfigReaderFailures, Config] = {
    for {
      _ <- invalidateCaches().right
      rawConfig <- load().right
      config <- improveFailures[Config](
        loadConfig[Config](rawConfig.getConfig(namespace))(conv), namespace, None).right
    } yield config
  }
```

but

```scala
def loadConfigOrThrow[Config](namespace: String)(implicit conv: ConfigConvert[Config], ct: ClassTag[Config]): Config = {
    val errorOrConfig =
      for {
        _ <- invalidateCaches().right
        rawConfig <- load().right.map(_.getConfig(namespace)).right
        config <- improveFailures[Config](
          loadConfig[Config](rawConfig)(conv), namespace, ConfigValueLocation(rawConfig.root())).right
      } yield config
    getResultOrThrow[Config](errorOrConfig)
  }
```

I'm not sure why the first set the position of the error to `None` while the second to `ConfigValueLocation(rawConfig.root())` but this is preventing me from reusing the first in the second. @jcazevedo do you know why the two methods are different for the error location?

leifwickland replied:
@melrief If you can't get resolution quickly, I'd be tempted to just revert 612b620 and pick up the thread in a separate issue focused on deduplication.

melrief replied:
I `git revert`ed the change with fadbf26 and I've opened #162 to work on the code reuse issue. Let's move the discussion in that issue.

jcazevedo replied:
> I'm not sure why the first set the position of the error to None while the second to ConfigValueLocation(rawConfig.root()) but this is preventing me from reusing the first in the second. @jcazevedo do you know why the two methods are different for the error location?

It looks like a mistake. I think the `location` on `loadConfig` should be obtained via `ConfigValueLocation(rawConfig.root())` as well.

melrief replied:
Then @leifwickland is right and we should always reuse `loadConfig` in `loadConfigOrThrow`. Let's tackle #162 in the near future and fix this inconsistency in the API.

jcazevedo replied:
Having had a better look at it, this is definitely a mistake. We should call `ConfigValueLocation(rawConfig.root())` [here](https://github.com/melrief/pureconfig/pull/161/files#diff-12918ae14aad73e551e678d26010a1deR46) and [here](https://github.com/melrief/pureconfig/pull/161/files#diff-12918ae14aad73e551e678d26010a1deR79) (and reuse this code on the `loadConfigOrThrow` methods).

I don't know if you prefer doing this in the context of this PR or in a separate one.

melrief replied:
Separate one but in this release. Let's do the review of this PR to manage exceptions from typesafe API and use the PR that solves #162 to fix the API. I prefer to have two commits in the git history tbh.

melrief committed:
reuse loadConfig for loadConfigOrThrow

leifwickland commented on core/src/main/scala/pureconfig/package.scala:
```diff
@@ -143,9 +166,7 @@ package object pureconfig {
    */
   @throws[ConfigReaderException[_]]
   def loadConfigOrThrow[Config](path: Path)(implicit conv: ConfigConvert[Config], ct: ClassTag[Config]): Config = {
```
I can't figure out where the `ClassTag` is getting used. `loadConfig` doesn't require it. I think it can go.

I think that's also true of all (or most) of the other `loadConfigOrThrow` methods.

melrief replied:
The `ClassTag` is propagated all the way down to the `ConfigReaderException`, which uses it to know the runtime name of the class for the error message.

melrief committed:
Revert "reuse loadConfig for loadConfigOrThrow"

This reverts commit 612b620107d8df4820494c4c6988e3a1814199cb.

jcazevedo commented on core/src/main/scala/pureconfig/backend/TypesafeConfigWrapper.scala:
```diff
@@ -0,0 +1,38 @@
+ * A wrapper of [[com.typesafe.config.ConfigFactory]] whose methods return `Either` instead
+ * of throwing exceptions
+ */
+object ConfigFactoryWrapper {
```
I prefer having this file named `ConfigFactoryWrapper`, to better reflect its contents.

melrief committed:
TypesafeConfigWrapper.scala => ConfigFactoryWrapper.scala

melrief commented:
@jcazevedo with 3a7ca16 I've changed the name of the file and added a custom failure called `CannotParse` for typesafe config parsing errors. I'd like to not add too many customs failures here because I think goal should be to be generic over the backend. Our failures should reflect a generic backend error rather than a specific one. Parsing errors, for instance, are generic but `NotResolved` seems specific to typesafe config.
I'm open to objections on this but I really like the idea ofallowing to choose the backend for parsing the configuration. It won't be easy but it's doable.
Let me know what do you think.

melrief committed:
add missing space in imports

jcazevedo commented:
> I'd like to not add too many customs failures here because I think goal should be to be generic over the backend. Our failures should reflect a generic backend error rather than a specific one. Parsing errors, for instance, are generic but NotResolved seems specific to typesafe config.

It makes perfect sense. 

This looks good to merge as it is. Thanks!
